### PR TITLE
Allow tx receipts to be retrieved via RPC

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

https://github.com/obscuronet/obscuro-internal/issues/479. Allows transaction receipts to be retrieved via RPC. The RPC requests and responses are encrypted using viewing keys.

### What changes were made as part of this PR:

Functional.

- Introduce RPC endpoint to retrieve transaction receipts
- Wallet extension test cases for transaction receipts
- Adds method in enclave to return transaction receipts given transaction hash

### CODE REVIEW

File: go/obscuronode/nodecommon/rpc/generated/enclave.proto [hunk 1]
```diff
@@ -44,7 +44,8 @@ service EnclaveProto {
   // SubmitTx - user transactions
   rpc SubmitTx(SubmitTxRequest) returns (SubmitTxResponse) {}
 
-  // Balance - returns the balance of an address with a block delay
+  // ExecuteOffChainTransaction - returns the result of executing the smart contract as a user, encrypted with the
```
Comment (joeldudley) [inline go/obscuronode/nodecommon/rpc/generated/enclave.proto:47]: The old description was a copy-paste mistake.

File: go/obscuronode/host/enclave_client.go [hunk 2]
```diff
@@ -33,6 +33,8 @@ type EnclaveRPCClient struct {
 	config      config.HostConfig
 }
 
+// TODO - Avoid panicking and return errors instead where appropriate.
```
Comment (joeldudley) [inline go/obscuronode/host/enclave_client.go:36]: We already do this now - some methods return errors.

File: go/obscuronode/enclave/enclave.go [hunk 3]
```diff
@@ -575,6 +575,46 @@ func (e *enclaveImpl) GetTransaction(txHash common.Hash) *nodecommon.L2Tx {
 	}
 }
 
+func (e *enclaveImpl) GetTransactionReceipt(encryptedParams nodecommon.EncryptedParams) (nodecommon.EncryptedResponse, error) {
```
Comment (tudor-malene) [inline go/obscuronode/enclave/enclave.go:578]: shouldn't this be in the interface as well?

File: go/obscuronode/enclave/enclave.go [hunk 4]
```diff
@@ -575,6 +575,46 @@ func (e *enclaveImpl) GetTransaction(txHash common.Hash) *nodecommon.L2Tx {
...
+	txReceipt, err := e.storage.GetTransactionReceipt(txHash)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve transaction receipt in eth_getTransactionReceipt request. Cause: %w", err)
+	}
+	txReceiptBytes, err := txReceipt.MarshalJSON()
+	if err != nil {
+		return nil, fmt.Errorf("could not marshall transaction receipt to JSON in eth_getTransactionReceipt request. Cause: %w", err)
+	}
+
+	tx, _, _, _, err := e.storage.GetTransaction(txHash) //nolint:dogsled
```
Comment (tudor-malene) [inline go/obscuronode/enclave/enclave.go:600]: I'd split this up into a few methods 
Quite hard to figure out what's going on here

File: go/obscuronode/nodecommon/types.go [hunk 5]
```diff
@@ -21,6 +21,7 @@ type (
 	EncryptedTransactions []byte // A blob of encrypted transactions, as they're stored in the rollup.
 	EncryptedParams       []byte // The params for an off-chain call or getBalance request, as a JSON list encrypted with the public key of the enclave.
```
Comment (tudor-malene) [inline go/obscuronode/nodecommon/types.go:22]: we should create some structs for each message that is being sent. We can marshall it to json, but we should have structs. Quite hard to follow without them.
Also, you could create "EncryptedXXX []byte" for each of them. This way we could follow the logic better.

File: go/obscuronode/enclave/enclave.go (hunk 4 above)
Comment (joeldudley) [reply go/obscuronode/enclave/enclave.go:600]: Thanks, done.

**Comment by joeldudley** [comment]
> As a functional comment, I'd suggest we use: the rpc GetTransactionReceipt from the simulation. In simulation/utils.go- getTransaction, we can call the rpc receipt as well and check that it is successful

@tudor-malene I'm not sure what you mean. `GetTransaction` is defined as below - there is no issuance of a transaction. Do you mean somewhere else perhaps?

```
func getTransaction(client obscuroclient.Client, hash common.Hash) *nodecommon.L2Tx {
	method := obscuroclient.RPCGetTransaction

	var result *nodecommon.L2Tx
	err := client.Call(&result, method, hash)
	if err != nil {
		panic(fmt.Errorf("simulation failed due to failed %s RPC call. Cause: %w", method, err))
	}

	return result
}
```

**Comment by tudor-malene** [comment]
> > As a functional comment, I'd suggest we use: the rpc GetTransactionReceipt from the simulation. In simulation/utils.go- getTransaction, we can call the rpc receipt as well and check that it is successful
> 
> @tudor-malene I'm not sure what you mean. `GetTransaction` is defined as below - there is no issuance of a transaction. Do you mean somewhere else perhaps?
> 
> ```
> func getTransaction(client obscuroclient.Client, hash common.Hash) *nodecommon.L2Tx {
> 	method := obscuroclient.RPCGetTransaction
> 
> 	var result *nodecommon.L2Tx
> 	err := client.Call(&result, method, hash)
> 	if err != nil {
> 		panic(fmt.Errorf("simulation failed due to failed %s RPC call. Cause: %w", method, err))
> 	}
> 
> 	return result
> }
> ```

it was just a suggestion to keep it simple.
in this `getTransaction` function we make an rpc call to get the actual transaction. I was suggesting to also make an rpc call to get the receipt.
This way we know each injected transaction has a stored receipt and the status is correct

File: go/obscuronode/nodecommon/types.go (hunk 5 above)
Comment (joeldudley) [reply go/obscuronode/nodecommon/types.go]: I think I've addressed this. Can you check?

File: go/obscuronode/enclave/enclave.go [hunk 6]
```diff
@@ -575,6 +575,34 @@ func (e *enclaveImpl) GetTransaction(txHash common.Hash) *nodecommon.L2Tx {
...
+	viewingKeyAddress, err := e.getTxSender(txHash)
+	if err != nil {
+		return nil, err
+	}
+
+	txReceipt, err := e.storage.GetTransactionReceipt(txHash)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve transaction receipt in eth_getTransactionReceipt request. Cause: %w", err)
+	}
+	txReceiptBytes, err := txReceipt.MarshalJSON()
```
Comment (tudor-malene) [inline go/obscuronode/enclave/enclave.go:593]: feels like marshaling belongs logically with the encryption.

**Review by tudor-malene — APPROVED** [review]
looks good. One very small stylistic comment and then good to merge.

File: go/obscuronode/enclave/enclave.go (hunk 6 above)
Comment (joeldudley) [reply go/obscuronode/enclave/enclave.go:593]: @tudor-malene If I do that, I need a separate encryption function for each type (balance, call, tx receipt).
